File: ui-info/src/main/java/l/files/ui/info/InfoFragment.java
package l.files.ui.info;

import android.graphics.Bitmap;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.util.DisplayMetrics;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import java.util.ArrayList;

import l.files.fs.Name;
import l.files.fs.Path;
import l.files.fs.Stat;
import l.files.ui.base.graphics.Rect;
import l.files.ui.preview.Preview;

import static android.text.format.DateUtils.FORMAT_SHOW_DATE;
import static android.text.format.DateUtils.FORMAT_SHOW_TIME;
import static android.text.format.DateUtils.formatDateTime;
import static android.view.View.GONE;
import static java.util.Collections.singleton;
import static java.util.concurrent.TimeUnit.MILLISECONDS;
import static l.files.base.Objects.requireNonNull;

public final class InfoFragment
        extends InfoBaseFragment
        implements Preview.Callback {

    private static final String ARG_STAT = "stat";

    public static InfoFragment create(Path path, @Nullable Stat stat) {
        requireNonNull(path);
        return newFragment(newArgs(path, stat));
    }

    private static Bundle newArgs(Path path, @Nullable Stat stat) {
        ArrayList<Name> names = new ArrayList<>(singleton(path.name()));
        Bundle bundle = new Bundle();
        bundle.putParcelableArrayList(ARG_CHILDREN, names);
        bundle.putParcelable(ARG_PARENT_DIRECTORY, path.parent());
        bundle.putParcelable(ARG_STAT, stat);
        return bundle;
    }

    private static InfoFragment newFragment(Bundle bundle) {
        InfoFragment fragment = new InfoFragment();
        fragment.setArguments(bundle);
        return fragment;
    }

    @Nullable
    private Stat stat;
    private Path path;

    private Rect constraint;
    private View backgroundView;
    private TextView nameView;
    private TextView lastModifiedView;
    private ImageView thumbnailView;

    public CharSequence getDisplayedName() {
        return nameView.getText();
    }

    public CharSequence getDisplayedLastModifiedTime() {
        return lastModifiedView.getText();
    }

    @Override
    int layoutResourceId() {
        return R.layout.info_fragment;
    }

    @Override
    public void onViewCreated(
            @NonNull View view,
            @Nullable Bundle savedInstanceState
    ) {
        super.onViewCreated(view, savedInstanceState);
        findViews(view);
        findArgs(getArguments());
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        updateViews();
    }

    private void findViews(View view) {
        backgroundView = view.findViewById(R.id.root);
        nameView = view.findViewById(R.id.name);
        thumbnailView = view.findViewById(R.id.image);
        lastModifiedView = view.findViewById(R.id.modified);
        constraint = calculateConstraint();
    }

    private void findArgs(Bundle args) {
        path = getParentDirectory().concat(getChildren().get(0));
        stat = args.getParcelable(ARG_STAT);
    }

    private void updateViews() {
        updateNameView(path);
        if (stat != null) {
            updateBackgroundView(path, stat);
            updateThumbnailView(path, stat);
            updateLastModifiedView(stat);
        }
    }

    private void updateNameView(Path path) {
        Name name = path.name();
        nameView.setMaxWidth(constraint.width());
        nameView.setText(name != null
                ? name.toString()
                : path.toString()
        );
    }

    @Override
    String formatSize(long size, int count) {
        if (stat == null || stat.isDirectory()) {
            return super.formatSize(size, count);
        } else {
            return formatSize(size);
        }
    }

    private void updateLastModifiedView(Stat stat) {
        long millis = stat.lastModifiedTime().to(MILLISECONDS);
        int flags = FORMAT_SHOW_DATE | FORMAT_SHOW_TIME;
        String text = formatDateTime(getActivity(), millis, flags);
        lastModifiedView.setText(text);
    }

    private Rect calculateConstraint() {
        DisplayMetrics metrics = getResources().getDisplayMetrics();
        int width = (int) (metrics.widthPixels * 0.75);
        int height = metrics.heightPixels;
        return Rect.of(width, height);
    }

    private void updateBackgroundView(Path file, Stat stat) {
        Preview preview = Preview.get(getActivity());
        Bitmap bg = preview.getBlurredThumbnail(file, stat, constraint, true);
        if (bg != null) {
            updateBackgroundView(bg);
        }
    }

    private void updateBackgroundView(Bitmap bitmap) {
        Drawable bg = new BitmapDrawable(getResources(), bitmap);
        bg.setAlpha((int) (0.3f * 255));
        backgroundView.setBackground(bg);
    }

    private void updateThumbnailView(Path file, Stat stat) {
        Preview preview = Preview.get(getActivity());

        Bitmap thumbnail = preview.getThumbnail(file, stat, constraint, true);
        if (thumbnail != null) {
            thumbnailView.setImageBitmap(thumbnail);
            return;
        }

        Rect size = preview.getSize(file, stat, constraint, false);
        if (size != null) {
            setImageViewMinSize(size);
        }
        preview.get(file, stat, constraint, this, getContext());

    }

    private Rect scaleSize(Rect size) {
        return size.scaleDown(constraint);
    }

    private void setImageViewMinSize(Rect size) {
        Rect scaled = scaleSize(size);
        thumbnailView.setMinimumWidth(scaled.width());
        thumbnailView.setMinimumHeight(scaled.height());
    }

    @Override
    public void onPreviewAvailable(Path file, Stat stat, Bitmap thumbnail) {
        showImageView(thumbnail);
    }

    private void showImageView(Bitmap thumbnail) {
        thumbnailView.setImageBitmap(thumbnail);
        thumbnailView.setAlpha(0F);
        thumbnailView.animate().alpha(1).setDuration(animationDuration());
    }

    private int animationDuration() {
        int id = android.R.integer.config_mediumAnimTime;
        return getResources().getInteger(id);
    }

    @Override
    public void onBlurredThumbnailAvailable(Path path, Stat stat, Bitmap bm) {
    }

    @Override
    public void onPreviewFailed(Path path, Stat stat, Object cause) {
        hindImageView();
    }

    private void hindImageView() {
        thumbnailView.setImageDrawable(null);
        thumbnailView.setVisibility(GONE);
    }

}

File: ui-info/src/main/java/l/files/ui/info/InfoFragment.kt
package l.files.ui.info

import android.graphics.Bitmap
import android.graphics.drawable.BitmapDrawable
import android.os.Bundle
import android.text.format.DateUtils.*
import android.view.View
import kotlinx.android.synthetic.main.info_fragment.*
import l.files.ui.base.graphics.Rect
import l.files.ui.preview.Preview
import l.files.ui.preview.getPreview
import java.nio.file.Path
import java.nio.file.attribute.BasicFileAttributes
import java.time.Instant

class InfoFragment : InfoBaseFragment(), Preview.Callback {

  private var time: Instant? = null
  private var isDirectory: Boolean? = null
  private lateinit var path: Path
  private lateinit var constraint: Rect

  val displayedName: CharSequence
    get() = nameView.text

  val displayedLastModifiedTime: CharSequence
    get() = lastModifiedView.text

  override fun layoutResourceId() = R.layout.info_fragment

  override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
    super.onViewCreated(view, savedInstanceState)
    constraint = calculateConstraint()
    path = readParentDirectory().resolve(readChildren().iterator().next())
    time = requireArguments().getSerializable(ARG_TIME) as Instant
    isDirectory = requireArguments().getSerializable(ARG_IS_DIR) as Boolean
  }

  override fun onActivityCreated(savedInstanceState: Bundle?) {
    super.onActivityCreated(savedInstanceState)
    updateViews()
  }

  private fun updateViews() {
    updateNameView(path)
    if (time != null) {
      updateBackgroundView(path, time!!)
      updateThumbnailView(path, time!!)
      updateLastModifiedView(time!!)
    }
  }

  private fun updateNameView(path: Path) {
    val name = path.fileName
    nameView.maxWidth = constraint.width()
    nameView.text = name?.toString() ?: path.toString()
  }

  override fun formatSize(size: Long, count: Int): String =
    if (isDirectory == null || isDirectory!!) {
      super.formatSize(size, count)
    } else {
      formatSize(size)
    }

  private fun updateLastModifiedView(time: Instant) {
    val millis = time.toEpochMilli()
    val flags = FORMAT_SHOW_DATE or FORMAT_SHOW_TIME
    val text = formatDateTime(activity, millis, flags)
    lastModifiedView.text = text
  }

  private fun calculateConstraint(): Rect {
    val metrics = resources.displayMetrics
    val width = (metrics.widthPixels * 0.75).toInt()
    val height = metrics.heightPixels
    return Rect.of(width, height)
  }

  private fun updateBackgroundView(path: Path, time: Instant) {
    val preview = requireContext().getPreview()
    val bg = preview.getBlurredThumbnail(path, time, constraint, true)
    bg?.let { updateBackgroundView(it) }
  }

  private fun updateBackgroundView(bitmap: Bitmap) {
    val bg = BitmapDrawable(resources, bitmap)
    bg.alpha = (0.3f * 255).toInt()
    backgroundView.background = bg
  }

  private fun updateThumbnailView(path: Path, time: Instant) {
    val context = requireContext()
    val preview = context.getPreview()
    val thumbnail =
      preview.getThumbnail(path, time, constraint, true)
    if (thumbnail != null) {
      thumbnailView.setImageBitmap(thumbnail)
      return
    }
    val size =
      preview.getSize(path, time, constraint, false)
    size?.let { setImageViewMinSize(it) }
    preview.get(path, time, constraint, this, context)
  }

  private fun scaleSize(size: Rect): Rect = size.scaleDown(constraint)

  private fun setImageViewMinSize(size: Rect) {
    val scaled = scaleSize(size)
    thumbnailView.minimumWidth = scaled.width()
    thumbnailView.minimumHeight = scaled.height()
  }

  override fun onPreviewAvailable(
    path: Path,
    time: Instant,
    thumbnail: Bitmap
  ) {
    showImageView(thumbnail)
  }

  private fun showImageView(thumbnail: Bitmap) {
    thumbnailView.setImageBitmap(thumbnail)
    thumbnailView.alpha = 0f
    thumbnailView.animate().alpha(1f).duration = animationDuration().toLong()
  }

  private fun animationDuration() =
    resources.getInteger(android.R.integer.config_mediumAnimTime)

  override fun onBlurredThumbnailAvailable(
    path: Path,
    time: Instant,
    thumbnail: Bitmap
  ) {
  }

  override fun onPreviewFailed(path: Path, time: Instant, cause: Any) =
    hindImageView()

  private fun hindImageView() {
    thumbnailView.setImageDrawable(null)
    thumbnailView.visibility = View.GONE
  }

  companion object {
    private const val ARG_TIME = "time"
    private const val ARG_IS_DIR = "is-dir"

    @JvmStatic
    fun create(path: Path, attrs: BasicFileAttributes?): InfoFragment =
      newFragment(newArgs(path, attrs))

    private fun newArgs(path: Path, attrs: BasicFileAttributes?): Bundle {
      val bundle = Bundle()
      bundle.putStringArrayList(
        ARG_CHILDREN,
        arrayListOf(path.fileName.toString())
      )
      bundle.putString(ARG_PARENT_DIRECTORY, path.parent.toString())
      bundle.putSerializable(ARG_TIME, attrs?.lastModifiedTime()?.toInstant())
      bundle.putSerializable(ARG_IS_DIR, attrs?.isDirectory)
      return bundle
    }

    private fun newFragment(bundle: Bundle): InfoFragment {
      val fragment = InfoFragment()
      fragment.arguments = bundle
      return fragment
    }
  }
}
